src/types/value.rs

```diff
@@ -0,0 +1,738 @@
+                    n.as_f64().unwrap().into_py(py)
+                } else {
+                    Option::<()>::None.into_py(py)
+                }
+            }
+            Value::Sequence(s) => {
+                let mut pyseq = vec![];
+                for v in s.iter() {
+                    pyseq.push(v.as_py_obj(py)?);
+                }
+                pyseq.into_py(py)
+            }
+            Value::Mapping(m) => m.as_py_dict(py)?.into(),
+            Value::Null => Option::<()>::None.into_py(py),
+            // ValueList should never get emitted to Python
+            Value::ValueList(_) => unreachable!(),
+        };
+        Ok(obj)
+    }
+
+    /// Handles special mapping key prefix values for `String` Values.
+    ///
+    /// For String values, if a mapping key prefix is present, the prefix is stripped from the
+    /// String, and the corresponding `KeyPrefix` variant is returned. Otherwise, the string is
+    /// cloned and returned.
+    ///
+    /// For non-String values, the value is unconditionally cloned and returned unmodified.
+    #[allow(unused)]
+    pub(super) fn strip_prefix(&self) -> (Self, Option<KeyPrefix>) {
+        match self {
+            Self::String(s) => {
+                if s.is_empty() {
+                    return (self.clone(), None);
+                }
+                let p = KeyPrefix::from(s.chars().next().unwrap());
+                if p.is_some() {
+                    (Self::String(s[1..].to_string()), p)
+                } else {
+                    (self.clone(), None)
+                }
+            }
+            _ => (self.clone(), None),
+        }
+    }
+}
+
+#[cfg(test)]
+mod value_tests {
+    use super::*;
+    use paste::paste;
+
+    #[test]
+    fn test_is_null() {
+        assert!(Value::Null.is_null());
+        assert!(!Value::Bool(true).is_null());
+    }
+
+    #[test]
+    fn test_is_bool() {
+        assert!(!Value::Null.is_bool());
+        assert!(Value::Bool(true).is_bool());
+    }
+
+    #[test]
+    fn test_as_bool() {
+        let b = Value::Bool(true);
+        assert_eq!(b.as_bool(), Some(true));
+        assert_eq!(Value::Null.as_bool(), None);
+    }
+
+    macro_rules! test_number {
+        ($($ty:ident $val:expr)*) => {
+            $(
+                paste! {
+                #[test]
+                fn [<test_is_ $ty>]() {
+                    assert!(!Value::Null.[<is_ $ty>]());
+                    let n: $ty = $val;
+                    let n = Value::Number(n.into());
+                    assert!(n.[<is_ $ty>]());
+                }
+
+                #[test]
+                fn [<test_as_ $ty>]() {
+                    assert_eq!(Value::Null.[<as_ $ty>](), None);
+                    let n: $ty = $val;
+                    let n = Value::Number(n.into());
+                    assert_eq!(n.[<as_ $ty>](), Some($val));
+                }
+                }
+            )*
+        }
+    }
+    test_number! { u64 5 i64 -3 f64 3.14 }
```

Comment:
side-note: Maybe we should use a crate to generate parameterized tests (I've used https://crates.io/crates/parameterized_test for a private project before) instead of parameterizing tests by hand using raw macros.

Reply:
might hide some complexities of creating your own macros

Reply:
I'll track that in the internal ticket for now, and will create a follow-up if I don't manage to set something up in this sprint.